src/refdb_fs.c

```diff
@@ -1375,12 +1422,17 @@ static int refdb_fs_backend__delete_tail(
 
 	git_sortedcache_wunlock(backend->refcache);
 
-	if (error == GIT_ENOTFOUND) {
-		error = loose_deleted ? 0 : ref_error_notfound(ref_name);
+	if (error != GIT_ENOTFOUND && (error = packed_write(backend)) < 0)
+		goto cleanup;
+	else if (error == GIT_ENOTFOUND && !is_loose) {
+		error = ref_error_notfound(ref_name);
 		goto cleanup;
 	}
 
-	error = packed_write(backend);
+	/* unlock here, so that we can delete the all empty parents */
+	git_filebuf_cleanup(file);
```

Comment:
We need to close the lock file before we can remove the possibly empty parents, otherwise it won't work. And we still need it in the cleanup path. As cleanup disposes the internal buffer, calling it twice doesn't cause problems.